src/Microsoft.DotNet.Interactive.CSharpProject/RoslynWorkspaceUtilities.cs

```diff
@@ -0,0 +1,389 @@
+        }
+
+        return workspace.CurrentSolution.GetProject(projectId);
+    }
+
+    internal static AdhocWorkspace CreateWorkspace()
+    {
+        AdhocWorkspace workspace = new AdhocWorkspace();
+        workspace.WorkspaceChanged += WorkspaceChangedHandler;
+        workspace.WorkspaceFailed += WorkspaceFailedHandler;
+        return workspace;
+    }
+
+    private static void WorkspaceFailedHandler(object sender, WorkspaceDiagnosticEventArgs e)
+    {
+        // Log error
+    }
+
+    private static void WorkspaceChangedHandler(object sender, WorkspaceChangeEventArgs e)
+    {
+        // Log
+    }
+
+    private static ProjectInfo GetProjectInfo(BuildProjectData buildProjectData, Microsoft.CodeAnalysis.Workspace workspace, ProjectId projectId)
+    {
+        string projectName = Path.GetFileNameWithoutExtension(buildProjectData.ProjectFilePath);
+        if (!TryGetSupportedLanguageName(buildProjectData.ProjectFilePath, out string languageName))
+        {
+            return null;
+        }
+        return ProjectInfo.Create(
+            projectId,
+            VersionStamp.Create(),
+            projectName,
+            projectName,
+            languageName,
+            filePath: buildProjectData.ProjectFilePath,
+            outputFilePath: buildProjectData.PropertyTargetPath,
+            documents: GetDocuments(buildProjectData, projectId),
+            projectReferences: GetExistingProjectReferences(buildProjectData, workspace),
+            metadataReferences: GetMetadataReferences(buildProjectData),
+            analyzerReferences: GetAnalyzerReferences(buildProjectData, workspace),
+            parseOptions: CreateParseOptions(buildProjectData, languageName),
+            compilationOptions: CreateCompilationOptions(buildProjectData, languageName));
+    }
+
+    private static bool TryGetSupportedLanguageName(string projectPath, out string languageName)
+    {
+        switch (Path.GetExtension(projectPath))
+        {
+            case ".csproj":
+                languageName = LanguageNames.CSharp;
+                return true;
+            default:
+                languageName = null;
+                return false;
+        }
+    }
+
+    private static IEnumerable<DocumentInfo> GetDocuments(BuildProjectData buildProjectData, ProjectId projectId) =>
+        buildProjectData
+            .SourceFiles?.Where(File.Exists)
+    .Select(x => DocumentInfo.Create(
+                DocumentId.CreateNewId(projectId),
+    Path.GetFileName(x),
+                loader: TextLoader.From(
+                    TextAndVersion.Create(
+                        SourceText.From(File.ReadAllText(x), Encoding.Unicode), VersionStamp.Create())),
+                filePath: x))
+        ?? Array.Empty<DocumentInfo>();
+
+    private static IEnumerable<ProjectReference> GetExistingProjectReferences(BuildProjectData buildProjectData, Microsoft.CodeAnalysis.Workspace workspace) =>
```

Comment:
Ditto.

Reply:
Done.

Now this method returns IReadOnlyList